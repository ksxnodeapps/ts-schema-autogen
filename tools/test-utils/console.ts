import { ConsoleInstance } from 'simple-fake-console'
import { Console } from '@ts-schema-autogen/types'

export class FakeConsole implements Console.Mod {
  public readonly core = new ConsoleInstance()
  public readonly info = jest.fn(this.core.info)
  public readonly error = jest.fn(this.core.error)
}
