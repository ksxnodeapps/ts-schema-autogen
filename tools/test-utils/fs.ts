import { StringPathFileSystem } from 'simple-fake-fs'
import { FSX } from '@ts-schema-autogen/types'

export class FakeFileSystem extends StringPathFileSystem implements FSX.Mod {
  private mkfn<Xs extends any[], Y> (fn: (...xs: Xs) => Y) {
    return jest.fn(async (...xs: Xs) => fn.apply(this, xs))
  }

  public readonly stat = this.mkfn(this.statSync)
  public readonly readdir = this.mkfn(this.readdirSync)
  public readonly readFile = this.mkfn(this.readFileSync)
  public readonly outputFile = this.mkfn(this.outputFileSync)
  public readonly writeFile = this.mkfn(this.writeFileSync)
  public readonly remove = jest.fn()
}
