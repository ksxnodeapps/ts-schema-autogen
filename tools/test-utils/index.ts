export * from './console'
export * from './fs'
export * from './path'
export * from './print-result'
export * from './process'
export * from './typescript-json-schema'
