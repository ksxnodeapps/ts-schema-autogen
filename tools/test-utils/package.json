{
  "name": "@tools/test-utils",
  "version": "0.0.0",
  "private": true,
  "dependencies": {
    "simple-fake-fs": "^0.0.8",
    "simple-fake-console": "^0.1.6",
    "simple-fake-path": "^0.0.2",
    "@ts-schema-autogen/status": "file:../../packages/status",
    "@ts-schema-autogen/types": "file:../../packages/types"
  }
}
