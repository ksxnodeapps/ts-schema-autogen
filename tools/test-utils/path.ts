import { Path } from '@ts-schema-autogen/types'

export class FakePath implements Path.Mod {
  constructor (private readonly cwd: string) {}

  public readonly join = (...paths: string[]): string => {
    if (!paths.length) return '.'
    const [head, ...rest] = paths
    const tail = this.join(...rest)
    if (head === '.') return tail
    if (tail === '.') return head
    if (tail.startsWith('../')) return this.join(this.dirname(head), tail.slice('../'.length))
    const left = head.endsWith('/') ? head.slice(0, -1) : head
    const right = tail.startsWith('/') ? tail.slice(1) : tail
    return left + '/' + right
  }

  public readonly resolve = (...paths: string[]): string => {
    if (!paths.length) return '.'
    const [head, ...rest] = paths
    const tail = this.resolve(...rest)
    if (tail.startsWith('/')) return tail
    return this.join(this.cwd, head, tail)
  }

  public readonly dirname = (path: string): string => {
    if (path === '' || path === '.' || path === '/') return path
    const segments = path.split('/').slice(0, -1)
    if (!segments.length) return '.'
    return this.join(...segments)
  }

  public readonly basename = (path: string): string => {
    const segments = path.split('/')
    return segments[segments.length - 1]
  }
}
