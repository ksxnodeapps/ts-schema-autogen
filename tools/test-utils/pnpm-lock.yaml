dependencies:
  '@ts-schema-autogen/types': 'link:../../packages/types'
  simple-fake-fs: 0.0.4
lockfileVersion: 5.1
packages:
  /@types/node/13.5.0:
    dev: false
    resolution:
      integrity: sha512-Onhn+z72D2O2Pb2ql2xukJ55rglumsVo1H6Fmyi8mlU9SvKdBk/pUSUAiBY/d9bAOF7VVWajX3sths/+g6ZiAQ==
  /simple-fake-fs/0.0.4:
    dependencies:
      '@types/node': 13.5.0
      tslib: 1.10.0
    dev: false
    resolution:
      integrity: sha512-GUwEIYXI6oSJdMZkP1hK3x084z6HX6uyHMrs3erJXEzdyMIrJ7HJEjlbtMrs6zsELvgx0wr7Y9zTK61+igHC8A==
  /tslib/1.10.0:
    dev: false
    resolution:
      integrity: sha512-qOebF53frne81cf0S9B41ByenJ3/IuH8yJKngAX35CmiZySA0khhkovshKK+jGCaMnVomla7gVlIcc3EvKPbTQ==
specifiers:
  '@ts-schema-autogen/types': ../../packages/types
  simple-fake-fs: ^0.0.4
