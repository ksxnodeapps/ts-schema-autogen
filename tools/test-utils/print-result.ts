import { inspect } from 'util'
import { Success, Failure } from '@ts-schema-autogen/status'

export function printFailure (result: Failure<unknown>): string {
  let text = '\n'
  result.print((...args) => text += args.join(' ') + '\n')
  return text
}

export const printResult = (result: Success<unknown> | Failure<unknown>): string =>
  result.code ? printFailure(result) : inspect(result)
