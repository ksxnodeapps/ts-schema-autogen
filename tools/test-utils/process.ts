import { Process } from '@ts-schema-autogen/types'

export class FakeProcess implements Process.Mod {
  constructor (private readonly workDir: string) {}
  public readonly cwd = jest.fn(() => this.workDir)
}
