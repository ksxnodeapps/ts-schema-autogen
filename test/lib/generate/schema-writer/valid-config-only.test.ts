import fsTree from '../../../fixtures/fs-tree/valid-config-only'
import TJS from '../../../fixtures/tjs'
import { FakeFileSystem, FakePath } from '@tools/test-utils'

import {
  Success,
  SchemaWriter,
  createJsonConfigParser,
  createYamlConfigParser
} from '@ts-schema-autogen/lib'

const getParsers = () => Promise.all([
  createJsonConfigParser('JSON Parser'),
  createYamlConfigParser('YAML Parser')
])

async function setup (configPaths: readonly string[]) {
  const fsx = new FakeFileSystem('/', fsTree)
  const path = new FakePath()
  const tjs = new TJS()
  const schemaWriter = new SchemaWriter({
    fsx,
    path,
    tjs,
    parsers: await getParsers()
  })
  const result = await schemaWriter.writeSchemas(configPaths)
  return { fsx, path, tjs, result }
}

describe('one config file that specifies one output file', () => {
  const configPaths = ['yaml/single-symbol/single-output/output-filename/.schema.autogen.yaml']
  const expectedOutput = 'yaml/single-symbol/single-output/output-filename/output.schema.json'

  it('returns a Success', async () => {
    const { result } = await setup(configPaths)
    expect(result).toBeInstanceOf(Success)
  })

  it('calls outputFile once', async () => {
    const { fsx } = await setup(configPaths)
    expect(fsx.outputFile).toBeCalledTimes(1)
  })

  it('calls outputFile with expected argument', async () => {
    const { fsx } = await setup(configPaths)
    expect(fsx.outputFile).toBeCalledWith(expectedOutput, expect.any(String))
  })

  it('creates output file', async () => {
    const { fsx } = await setup(configPaths)
    const text = fsx.readFileSync(expectedOutput)
    const snapshot = JSON.parse(text)
    expect(snapshot).toMatchSnapshot()
  })
})

describe('multiple config files and multiple output', () => {
  async function getConfigPaths () {
    const path = new FakePath()
    const { product, map } = await import('iter-tools')
    const base = () => product(
      ['single-symbol', 'multiple-symbol'],
      ['single-output', 'multiple-output'],
      ['output-filename', 'output-descriptor']
    )
    const yaml = map(
      prefix => path.join('yaml', ...prefix, '.schema.autogen.yaml'),
      base()
    )
    const json = map(
      prefix => path.join('json', ...prefix, '.schema.autogen.json'),
      base()
    )
    return [...yaml, ...json]
  }

  it('returns a Success', async () => {
    const { result } = await setup(await getConfigPaths())
    expect(result).toBeInstanceOf(Success)
  })

  it('calls outputFile', async () => {
    const { fsx } = await setup(await getConfigPaths())
    const snapshot = fsx.outputFile.mock.calls
      .map(([filename, content]) => ({ filename, content: JSON.parse(content) }))
    expect(snapshot).toMatchSnapshot()
  })
})
