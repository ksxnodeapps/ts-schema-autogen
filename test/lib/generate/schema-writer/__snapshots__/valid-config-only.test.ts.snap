// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiple config files and multiple output calls outputFile 1`] = `
Array [
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "yaml/single-symbol/single-output/output-filename/output.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "yaml/single-symbol/single-output/output-descriptor/output.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "yaml/single-symbol/multiple-output/output-filename/abc.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "yaml/single-symbol/multiple-output/output-filename/def.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "yaml/single-symbol/multiple-output/output-filename/ghi.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "yaml/single-symbol/multiple-output/output-descriptor/non-descriptor.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "yaml/single-symbol/multiple-output/output-descriptor/default-descriptor.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "yaml/single-symbol/multiple-output/output-descriptor/custom-descriptor.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "yaml/multiple-symbol/single-output/output-filename/foo.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Bar",
      "type": "schema",
    },
    "filename": "yaml/multiple-symbol/single-output/output-filename/bar.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "yaml/multiple-symbol/single-output/output-descriptor/foo.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Bar",
      "type": "schema",
    },
    "filename": "yaml/multiple-symbol/single-output/output-descriptor/bar.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "yaml/multiple-symbol/multiple-output/output-filename/abc.foo.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "yaml/multiple-symbol/multiple-output/output-filename/def.foo.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "yaml/multiple-symbol/multiple-output/output-filename/ghi.foo.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Bar",
      "type": "schema",
    },
    "filename": "yaml/multiple-symbol/multiple-output/output-filename/abc.bar.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Bar",
      "type": "schema",
    },
    "filename": "yaml/multiple-symbol/multiple-output/output-filename/def.bar.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Bar",
      "type": "schema",
    },
    "filename": "yaml/multiple-symbol/multiple-output/output-filename/ghi.bar.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "yaml/multiple-symbol/multiple-output/output-descriptor/foo.filename.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "yaml/multiple-symbol/multiple-output/output-descriptor/foo.default.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "yaml/multiple-symbol/multiple-output/output-descriptor/foo.custom.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Bar",
      "type": "schema",
    },
    "filename": "yaml/multiple-symbol/multiple-output/output-descriptor/bar.4.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "json/single-symbol/single-output/output-filename/output.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "json/single-symbol/single-output/output-descriptor/output.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "json/single-symbol/multiple-output/output-filename/abc.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "json/single-symbol/multiple-output/output-filename/def.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "json/single-symbol/multiple-output/output-filename/ghi.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "json/single-symbol/multiple-output/output-descriptor/non-descriptor.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "json/single-symbol/multiple-output/output-descriptor/default-descriptor.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "json/single-symbol/multiple-output/output-descriptor/custom-descriptor.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "json/multiple-symbol/single-output/output-filename/foo.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Bar",
      "type": "schema",
    },
    "filename": "json/multiple-symbol/single-output/output-filename/bar.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "json/multiple-symbol/single-output/output-descriptor/foo.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Bar",
      "type": "schema",
    },
    "filename": "json/multiple-symbol/single-output/output-descriptor/bar.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "json/multiple-symbol/multiple-output/output-filename/abc.foo.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "json/multiple-symbol/multiple-output/output-filename/def.foo.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "json/multiple-symbol/multiple-output/output-filename/ghi.foo.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Bar",
      "type": "schema",
    },
    "filename": "json/multiple-symbol/multiple-output/output-filename/abc.bar.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Bar",
      "type": "schema",
    },
    "filename": "json/multiple-symbol/multiple-output/output-filename/def.bar.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Bar",
      "type": "schema",
    },
    "filename": "json/multiple-symbol/multiple-output/output-filename/ghi.bar.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "json/multiple-symbol/multiple-output/output-descriptor/foo.filename.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "json/multiple-symbol/multiple-output/output-descriptor/foo.default.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "json/multiple-symbol/multiple-output/output-descriptor/foo.custom.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Bar",
      "type": "schema",
    },
    "filename": "json/multiple-symbol/multiple-output/output-descriptor/bar.4.schema.json",
  },
]
`;

exports[`one config file that specifies multiple output file calls outputFile 1`] = `
Array [
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "yaml/multiple-symbol/multiple-output/output-descriptor/foo.filename.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "yaml/multiple-symbol/multiple-output/output-descriptor/foo.default.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Foo",
      "type": "schema",
    },
    "filename": "yaml/multiple-symbol/multiple-output/output-descriptor/foo.custom.schema.json",
  },
  Object {
    "content": Object {
      "symbol": "Bar",
      "type": "schema",
    },
    "filename": "yaml/multiple-symbol/multiple-output/output-descriptor/bar.4.schema.json",
  },
]
`;

exports[`one config file that specifies one output file creates output file 1`] = `
Object {
  "symbol": "Foo",
  "type": "schema",
}
`;
